### PR TITLE
{Core} az login: Capture managed_by_tenants error

### PR DESCRIPTION
**Pull request by jiasli**
**Description**  

If profile `latest` is used for Azure Stack cloud, `az login` will fail:

```powershell
> az cloud register -n redmond --endpoint-resource-manager "https://management.redmond.azurestack.corp.microsoft.com/" 
> az cloud set -n redmond
> az login
...
  File "d:\cli\azure-cli\src\azure-cli-core\azure\cli\core\_profile.py", line 276, in _normalize_properties
    subscription_dict[_MANAGED_BY_TENANTS] = [{_TENANT_ID: t.tenant_id} for t in s.managed_by_tenants]
TypeError: 'NoneType' object is not iterable
```

This is because `latest` should not be accepted by the service, and the service should return `Not Supported`. On the contrary, the service does accept the request and returns an invalid JSON which lacks `managedByTenants`. This behavior doesn't comply with the spec (https://github.com/Azure/azure-rest-api-specs/issues/9567), causing the parsing logic of CLI to fail.

Ideally, the user should run `az cloud set/update -n redmond --profile 2019-03-01-hybrid` to set the profile for the cloud before `az login`. 

This PR wraps the raw exception with better instructions:

```
UserFault: Invalid profile is used for cloud 'redmond'. To configure the cloud profile, 
run `az cloud set --name redmond --profile <profile>(e.g. 2019-03-01-hybrid)`. For more 
information about using Azure CLI with Azure Stack, see https://docs.microsoft.com/en-us
/azure/azure-stack/user/azure-stack-version-profiles-azurecli2#connect-to-azure-stack
```

### CODE REVIEW

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 1]
```diff
@@ -273,6 +273,16 @@ def _normalize_properties(self, user, subscriptions, is_service_principal, cert_
                 if hasattr(s, 'home_tenant_id'):
                     subscription_dict[_HOME_TENANT_ID] = s.home_tenant_id
                 if hasattr(s, 'managed_by_tenants'):
+                    if s.managed_by_tenants is None:
+                        # managedByTenants is missing from the response. This is a known service issue:
+                        # https://github.com/Azure/azure-rest-api-specs/issues/9567
+                        # pylint: disable=line-too-long
+                        from azure.cli.core.azclierror import UserFault
+                        raise UserFault("Invalid profile is used for cloud '{cloud_name}'. "
```
Comment (houk-ms) [inline src/azure-cli-core/azure/cli/core/_profile.py:281]: Can we define an error sub-class may be InvalidProfileError(UserFault) and raise the specific error? UserFault is designed as an abstract class and is not recommended to be raised directly.

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/_profile.py:281]: This is only a temporary mitigation to the service bug, it should actually be an `Unsupported API version` error.

**Comment by yonzhan** [comment]
Core

**Review by arrownj — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/_profile.py (hunk 1 above)
Comment (arrownj) [reply src/azure-cli-core/azure/cli/core/_profile.py:281]: @houk-ms , you can use `ABC` and `abstractmethod` (`from abc import ABC, abstractmethod`) to define your class and make the parent class can't be instantiated.

Comment (houk-ms) [reply src/azure-cli-core/azure/cli/core/_profile.py:281]: @arrownj sure, we can do that.

**Review by Juliehzl — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 2]
```diff
@@ -273,6 +273,16 @@ def _normalize_properties(self, user, subscriptions, is_service_principal, cert_
...
+                                        "To configure the cloud profile, run `az cloud set --name {cloud_name} --profile <profile>(e.g. 2019-03-01-hybrid)`. "
+                                        "For more information about using Azure CLI with Azure Stack, see "
+                                        "https://docs.microsoft.com/en-us/azure/azure-stack/user/azure-stack-version-profiles-azurecli2#connect-to-azure-stack"
```
Comment (fengzhou-msft) [inline src/azure-cli-core/azure/cli/core/_profile.py:284]:
```suggestion
                                        "https://docs.microsoft.com/azure/azure-stack/user/azure-stack-version-profiles-azurecli2#connect-to-azure-stack"
```

File: src/azure-cli-core/azure/cli/core/_profile.py (hunk 1 above)
Comment (houk-ms) [reply src/azure-cli-core/azure/cli/core/_profile.py:281]: @jiasli In this case, cosidering use `from azure.cli.core.azclierror import UnknownError` to replace UserFault if you don't define a new error type

**Review by fengzhou-msft — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/_profile.py (hunk 1 above)
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/_profile.py:281]:
```suggestion
                        from azure.cli.core.azclierror import UnknownError
                        raise UnknownError("Invalid profile is used for cloud '{cloud_name}'. "
```

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/_profile.py]: Let's use `CLIError` for now until the error can be classified as `UnsupportedAPIError` after the service is fixed.

**Comment by jiasli** [comment]
This PR has been shipped in Azure CLI 2.15.0.

**Pull request merged**